reader_core/src/gen7/reader.rs

```diff
@@ -151,32 +258,73 @@ impl Gen7Reader {
         hook::sos_seed()
     }
 
+    // We still use this naive check to hint to the RNG wrapper when it can rest.
+    pub fn sos_index(&self) -> u16 {
+        let index = pnp::read(self.addrs.sos_index);
+        if index != 624 { index } else { 0 }
+    }
+
+    pub fn sos_state(&self) -> u32 {
+        pnp::read(self.addrs.sos_sfmt_state + (self.sos_index() as u32 * 4))
+    }
+
+    /* Note: Not very useful...
+     * Value does not update until after last
+     * input, which makes it rather misleading. */
+    pub fn _sos_prevcall(&self) -> bool {
+        pnp::read_bool(self.addrs._sos_call_succeed)
+    }
+
     pub fn sos_chain(&self) -> u8 {
-        pnp::read(self.addrs.sos_chain_length)
+        pnp::read(self.addrs.sos_chain_len)
     }
+
     pub fn orb_active(&self) -> bool {
-        ((pnp::read::<u8>(self.addrs.orb_active) & 0x1) > 0) as bool
+        pnp::read_bool(self.addrs.orb_active)
     }
-    pub fn ally_slot(&self, caller_slot: u32, correction: u32) -> u32 {
-        if self.sos_chain() == 0 {
-            0
-        } else {
-            ((caller_slot - 1) + ((self.sos_chain() as i32 - (correction as i32 + 1)) % 3) as u32 + 1) % 4
+
+    pub fn wild_slot_lookup(&self, side: Gen7WildSide) -> Gen7PkmSlot {
+        match side.offset() {
+            Some(offset) => {
+                let battle_table = self.addrs.sos_battle_table + offset;
+                let pkx_container_ptr = pnp::read::<u32>(battle_table).clamp(
+                    self.addrs.pkm_container_base,
+                    self.addrs.pkm_container_base + (PKM_CONT_SIZE * 6),
+                );
+                Gen7PkmSlot::new(
+                    (((pnp::read::<u32>(pkx_container_ptr) + 0x40 - self.addrs.wild) / PK7_DATA_SIZE) + 1)
+                        .clamp(1, 6) as usize,
+                )
+            }
+            None => Gen7PkmSlot::Invalid,
+        }
+    }
```

Comment:
This all seems quite a bit more complicated than what currently exists.

If memory safety is the goal, I've had some other ideas for how I'd like to tackle that, and it might help improve readability.